src/target/ext/host_io.rs

```diff
@@ -362,11 +338,7 @@ pub trait HostIoReadlink: HostIo {
     /// will consume the `output` object and return a [`HostIoToken`]. This
     /// token ensures that the implementer of this method calls
     /// [`HostIoOutput::write`].
-    fn readlink<'a>(
-        &mut self,
-        filename: &[u8],
-        output: HostIoOutput<'a>,
-    ) -> HostIoResult<HostIoToken<'a>, Self>;
+    fn readlink<'a>(&mut self, filename: &[u8], buf: &mut [u8]) -> HostIoResult<usize, Self>;
```

Comment:
interesting... it seems like in this case, it would actually be _impossible_ to re-use the entire `PacketBuf` to send the response. This implies that #70 might not be possible for all cases.

Reply:
It may be not impossible with the current code structure, but we can always find a point where the data in `buf` is used and we don't start to use `buf` to prepare the response, so it's still possible?

If we are allowed to use multiple mutable reference ( like C ), we can assure it will safe to do it?

Reply:
I'm not entirely sure what you mean... but the point I was trying to make is that so long as the API signature has a `filename: &[u8]` that points somewhere into the middle of the PacketBuf, `buf: &mut [u8]` cannot point to the start of the packet buf, as you would have an aliased mutable pointer, which would be _very bad_ in Rust.

Off the top of my head, if we _really_ wanted to enable this use-case, we could do something like:

```rust
struct ReadLinkPacket<'a> {
    filename_start: usize,
    filename_end: usize,
    buf: &'a mut [u8]
}

impl<'a> ReadLinkPacket<'a> {
    fn filename(&self) -> &[u8] {
        &self.buf[self.filename_start..self.filename_end]
    }

    fn into_buf(self) -> &'a mut [u8] {
        self.buf
    }
}

fn readlink(&mut self, filename_and_buf: ReadLinkPacket<'_>) -> HostIoResult<usize, Self> {
    let filename = filename_and_buf.filename();
    let link = std::fs::read_link(filename) // etc...
    let buf = filename_and_buf.into_buf();
    // compose response using buf
}
```

Also, while thinking about this, I stumbled across an interesting observation: if you check the man page for the underlying `readlink` syscall, you'll find that the filename buffer and the output buffer are actually marked `restrict`, which means it wouldn't be possible to read the resolved link path directly into the buffer containing the filename. i.e: you _must_ either allocate a new buffer to read the resolved path into, or use memory disjoint from the `filename` buf, regardless if you were using Rust or C.

FWIW, I think the current implementation that doesn't use 100% of the PacketBuf is totally reasonable, especially since the underlying packet doesn't even have a mechanism of relaying "partial reads", and assumes that the packetbuf has enough space to fit the response.

Reply:
Let's take a look at how GDB implements it:
```C
#if !defined (PATH_MAX) || (PATH_MAX > (PBUFSIZ / 2 + 1))
#  define HOSTIO_PATH_MAX (PBUFSIZ / 2 + 1)
#else
#  define HOSTIO_PATH_MAX PATH_MAX
#endif
```
```C
static void
handle_readlink (char *own_buf, int *new_packet_len)
{
  char filename[HOSTIO_PATH_MAX], linkname[HOSTIO_PATH_MAX];
  char *p;
  int ret, bytes_sent;

  p = own_buf + strlen ("vFile:readlink:");

  if (require_filename (&p, filename)
      || require_end (p))
    {
      hostio_packet_error (own_buf);
      return;
    }

  if (hostio_fs_pid != 0 && the_target->multifs_readlink != NULL)
    ret = the_target->multifs_readlink (hostio_fs_pid, filename,
					linkname,
					sizeof (linkname) - 1);
  else
    ret = readlink (filename, linkname, sizeof (linkname) - 1);

  if (ret == -1)
    {
      hostio_error (own_buf);
      return;
    }

  bytes_sent = hostio_reply_with_data (own_buf, linkname, ret, new_packet_len);

  /* If the response does not fit into a single packet, do not attempt
     to return a partial response, but simply fail.  */
  if (bytes_sent < ret)
    sprintf (own_buf, "F-1,%x", FILEIO_ENAMETOOLONG);
}
```
So we have two ways to implement it:
1. Copy the filename into a temporary buf allocated on the stack and pass this buf to `ops.readlink`.
2. Allocate two global buf with the same size: `PacketBufIn` and `PacketBufOut`. This will cost twice of the memory, but the implementation will be very easy.

Reply:
> 1. Copy the filename into a temporary buf allocated on the stack and pass this buf to `ops.readlink`.

This forces `gdbstub` to arbitrarily decide what a reasonable "max path" size would be, which is something I've avoided. This is actually one of those cases where a callback-based output function would be ideal, since it gives the calling code the ability to allocate a stack-allocated buffer of whatever size it deems appropriate, and pass it back into `gdbstub` code.



> 2. Allocate two global buf with the same size: `PacketBufIn` and `PacketBufOut`. This will cost twice of the memory, but the implementation will be very easy.

I feel like this is something I brought up in the past (or maybe I just thought of it and didn't write it down), but yet, this would be the easiest way to handle these sorts of cases. The `gdbstub` build could be modified to accept an optional `PacketBufOut`, and the handler code would check for it's existence when deciding which `PacketBuf` to pass to the target implementation.

Reply:
> This forces `gdbstub` to arbitrarily decide what a reasonable "max path" size would be, which is something I've avoided. This is actually one of those cases where a callback-based output function would be ideal, since it gives the calling code the ability to allocate a stack-allocated buffer of whatever size it deems appropriate, and pass it back into `gdbstub` code.

What I said don't need `gdbstub` to decide "max path" size, because we already know the real length of `cmd.filename`, and allocate the buffer as needed.

> I feel like this is something I brought up in the past (or maybe I just thought of it and didn't write it down), but yet, this would be the easiest way to handle these sorts of cases. The `gdbstub` build could be modified to accept an optional `PacketBufOut`, and the handler code would check for it's existence when deciding which `PacketBuf` to pass to the target implementation.

It will be a mess. We pass entire `PacketBuf` to avoid possibility of overflow, if you don't know if the buffer will lead to overflow, how to write the code...
I think we can just split the `PacketBuf` user provided into two, and set `PacketSize` to half of the length of `PacketBuf`.

Reply:
> What I said don't need `gdbstub` to decide "max path" size, because we already know the real length of `cmd.filename`, and allocate the buffer as needed.

Maybe I misunderstood what you meant... could you sketch out a rough version of the code you're envisioning?

> I think we can just split the `PacketBuf` user provided into two, and set `PacketSize` to half of the length of `PacketBuf`.

The issue with this approach (or my approach for that matter), its that almost every other packet doesn't have this limitation, which means that you'd just end allocating a bunch of memory you're not using all that often.

Admittedly, "a bunch of memory" isn't really appropriate when we're talking about a 4k buffer on a PC with gigabytes of RAM, but if you consider `gdbstub` running on a embedded platform, we can't just waste memory like that willy-nilly (hence my reasoning for making it an explicitly optional bit of functionality, which individual implementations can opt into as part of the `gdbstub` constructor). Obviously, it's more messy, but that's the price you pay for modularity ¯\\\_(ツ)_/¯

Reply:
It's just:
```rust
HostIo::vFileReadlink(cmd) if ops.enable_readlink().is_some() => {
    let ops = ops.enable_readlink().unwrap();
    let mut stack_buf = [0; cmd.filename.len()];
    stack_buf.copy_from_slice(cmd.filename)
    handle_hostio_result! {
        if let Ok(ret) = ops.readlink(stack_buf, cmd.buf) => {
            res.write_str("F")?;
            res.write_num(ret)?;
            res.write_str(";")?;
            res.write_binary(cmd.buf.get(..ret).ok_or(Error::PacketBufferOverflow)?)?;
        }
    };
    HandlerStatus::Handled
}
```
Or if we choose `ReadLinkPacket` to solve multiple mutable reference problem:
```rust
let filename = filename_and_buf.filename();
let mut stack_buf = [0; filename.len()];
stack_buf.copy_from_slice(cmd.filename)
let buf = filename_and_buf.into_buf();
handle_hostio_result! {
    if let Ok(ret) = ops.readlink(stack_buf, buf) => {
        ......
    }
};
```
`ReadLinkPacket` is too ugly that it's impossible for us to use it as an argument of API, but we can use it in `gdbstub` internal code.

Reply:
Ahh, I see.

Unfortunately, `[0; cmd.filename.len()]` isn't valid Rust, since Rust doesn't support variable-length stack-allocated arrays (i.e: `alloca`). I'll admit, this is one of those features that I do hope Rust gains at some point, as while it's certainly dangerous and prone to misuse, it's a powerful tool for this kind of resource constrained systems programming...

This ties back to my earlier comment about having to decide on an "arbitrary max length", since the length of this stack allocated buffer would have to be a constant value. With sufficient plumbing, we could make this a bit "smarter" by making the `gdbstub` constructor take a const-generic value for the PacketBuf len, and then use that value as a stack allocated buf, but that also seems like a bad idea (i.e: `gdbstub` can be driven from IRQ contexts, with very limited stack space).

To make a long story short: it seems that the "solution" to using 100% of the PacketBuf in these sorts of cases will be to either do some serious API contortions (e.g: using that funky `ReadLinkPacket` code I sketched out earlier), or to throw our hands in the air, and require that users provide their own outgoing packet buffer for these sorts of cases.

Reply:
I think the key meta-insight with this particular packet is that it seems to require that _someone_ allocate an extra buffer _at some point_, either to store a copy of the provided `filename` (as you're suggesting), or to compose a response (as I'm suggesting). 

The reference implementation went with the first approach, where they picked an "arbitrary" max filename (based on either a user-defined `define`, or half the `PacketBuf` len) and then allocated it themselves. I'm not sure if I'm a huge fan of this approach from a "philosophically elegant code" POV, but it certainly seems like a reasonable "engineering realities" solution to the problem.

In `gdbstub`, I try my hardest to avoid these sorts of "arbitrary internal constants", and offer users the choice of how they'd like to handle these sorts of cases - either via a more complex API (e.g: using callbacks, type-shenanigans, etc...), or via internal switch statements (i.e: "give me another buffer as part of the constructor, and I'll use that one instead of the remaining bit of the packetbuf").

IMO, I think the current implementation is perfectly reasonable. We don't specify an arbitrary max file path, and if a future user runs into this limitation, I'd imagine one of two things will happen:

1. they'll just bump PacketSize up to compensate
2. They'll file an issue, at which point we can review their particular limitations, and decide on an approach that works for them

Reply:
> we could make this a bit "smarter" by making the `gdbstub` constructor take a const-generic value for the PacketBuf len, and then use that value as a stack allocated buf, but that also seems like a bad idea (i.e: `gdbstub` can be driven from IRQ contexts, with very limited stack space).

In such situation the user won't implement Host I/O, so we don't need to consider it?

> either do some serious API contortions (e.g: using that funky `ReadLinkPacket` code I sketched out earlier)

Didn't you say that only `ReadLinkPacket` is not enough, we still need a extra buffer?

Reply:
Yes, precisely: the meta argument I'm trying to make here is that it might not _matter_ that there are certain degenerate cases where this "partial buffer" approach will fail, since those cases are only likely to arise when running on a resource constrained platform where Host IO likely isn't being implemented.

> Didn't you say that only `ReadLinkPacket` is not enough, we still need a extra buffer?

Indeed, except this extra buffer would be allocated by the implementation itself (i.e: as part of the Target's handler implementation), as opposed to internally within `gdbstub`.

Reply:
> With sufficient plumbing, we could make this a bit "smarter" by making the `gdbstub` constructor take a const-generic value for the PacketBuf len, and then use that value as a stack allocated buf

Then I think it's not a bad idea? We can implement this way.

Reply:
As I mentioned in https://github.com/daniel5151/gdbstub/pull/72#issuecomment-922603362, I would like to shelve this discussion for now, and if you feel strongly about this limitation, feel free to open a new tracking issue for it, linking back to this thread + summarizing the various options both you and I have proposed.

Reply:
i.e: please give this comment a thumbs up if you're good to merge the PR as is, and I'll give it one final review to approve the code, at which point we'll merge this into `dev/0.6`.

Reply:
I think we can continue this discussion in #70?

Reply:
Ahh, yes, I am a silly man who is reviewing code far too late at night. Yes, we can certainly just continue discussing this at #70